promql/parse.go

```diff
@@ -0,0 +1,868 @@
+
+	p.expect(itemRightBracket, ctx)
+
+	// Parse optional offset.
+	if p.peek().typ == itemOffset {
+		p.next()
+		offi := p.expect(itemDuration, ctx)
+
+		offset, err = parseDuration(offi.val)
+		if err != nil {
+			p.error(err)
+		}
+	}
+
+	e := &MatrixSelector{
+		Name:          vs.Name,
+		LabelMatchers: vs.LabelMatchers,
+		Range:         erange,
+		Offset:        offset,
+	}
+	return e
+}
+
+// primaryExpr parses a primary expression.
+//
+//		<metric_namer> | <function_call> | <vector_aggregation> | <literal>
+//
+func (p *parser) primaryExpr() Expr {
+	switch t := p.next(); {
+	case t.typ == itemNumber:
+		n, err := strconv.ParseInt(t.val, 0, 64)
+		f := float64(n)
+		if err != nil {
+			f, err = strconv.ParseFloat(t.val, 64)
+		}
+		// TODO: Assuming infinity once we go over 1 << 64 seems intense.
+		if err != nil && err.(*strconv.NumError).Err != strconv.ErrRange {
+			p.errorf("error parsing number: %s", err)
+		}
+		return &NumberLiteral{clientmodel.SampleValue(f)}
+
+	case t.typ == itemString:
+		s := t.val[1 : len(t.val)-1]
+		return &StringLiteral{s}
+
+	case t.typ == itemLeftBrace:
+		// Metric selector without metric name.
+		p.backup()
+		return p.vectorSelector("")
+
+	case t.typ == itemIdentifier:
+		// Check for function call.
+		if p.peek().typ == itemLeftParen {
+			return p.call(t.val)
+		}
+		fallthrough // Else metric selector.
+
+	case t.typ == itemMetricIdentifier:
+		return p.vectorSelector(t.val)
+
+	case t.typ.isAggregator():
+		p.backup()
+		return p.aggrExpr()
+	}
+	p.errorf("invalid primary expression")
+	return nil
+}
+
+// labels parses a list of labelnames.
+//
+//		'(' <label_name>, ... ')'
+//
+func (p *parser) labels() clientmodel.LabelNames {
+	const ctx = "grouping opts"
+
+	p.expect(itemLeftParen, ctx)
+
+	labels := clientmodel.LabelNames{}
+	for {
+		id := p.expect(itemIdentifier, ctx)
+		labels = append(labels, clientmodel.LabelName(id.val))
+
+		if p.peek().typ != itemComma {
+			break
+		}
+		p.next()
+	}
+	p.expect(itemRightParen, ctx)
+
+	return labels
+}
+
+// aggrExpr parses an aggregation expression.
+//
+//		<aggr_op> (<vector_expr>) [by <labels>] [keeping_extra]
+//		<aggr_op> [by <labels>] [keeping_extra] (<vector_expr>)
+//
+func (p *parser) aggrExpr() *AggregateExpr {
+	const ctx = "aggregation"
+
+	agop := p.next()
+	if !agop.typ.isAggregator() {
+		p.errorf("%s is not an aggregation operator", agop)
+	}
+	var grouping clientmodel.LabelNames
+	var keepExtra bool
+
+	firstSyntax := false
+
+	if p.peek().typ == itemBy {
+		p.next()
+		grouping = p.labels()
+		firstSyntax = true
+	}
+	if p.peek().typ == itemKeepingExtra {
+		p.next()
+		keepExtra = true
+		firstSyntax = true
+	}
+
+	p.expect(itemLeftParen, ctx)
+	e := p.expr()
+	p.expect(itemRightParen, ctx)
+
+	if !firstSyntax {
+		if p.peek().typ == itemBy {
+			if len(grouping) > 0 {
+				p.errorf("aggregation must only contain one grouping clause")
+			}
+			p.next()
+			grouping = p.labels()
+		}
+		if p.peek().typ == itemKeepingExtra {
```

Comment:
If `keeping_extra` was used before, `firstSyntax` is true and this will not be reached.